anesthetic/utils.py

```diff
@@ -32,21 +32,66 @@ def logsumexp(a, axis=None, b=None, keepdims=False, return_sign=False):
                              return_sign=return_sign)
 
 
-def channel_capacity(w):
-    r"""Channel capacity (effective sample size).
+def effective_samples(w, gamma=1, method=None):
+    r"""Calculate effective number of samples.
 
-    .. math::
+    Using the Hugins-Roy family of effective samples
+    (https://aakinshin.net/posts/huggins-roy-ess/).
+
+    Parameters
+    ----------
+    gamma : int, float, default = 1
+        The value of gamma used to calculate the number of effective samples
+        according to
+
+        .. math::
+
+            n_{eff} = \bigg(\sum_{i=0}^n w^\gamma_i \bigg)^{\frac{1}{1-\beta}}
+
+        A value of 1 corresponds to the channel capacity or entropy based
+        calculation, 2 corresponds to a Kish estimate and gamma = 1/2
+        corresponds to a conservative estimate of the effective number of
+        samples. gamma has to be positive but can take any value.
 
-        H = \sum_i p_i \ln p_i
+    method : string
+        Method can be either 'channel' for a channel capacity estimate of
+        the effective number of samples or 'kish' for a Kish estimate
+        (Kish, Leslie (1965). Survey Sampling.
+        New York: John Wiley & Sons, Inc. ISBN 0-471-10949-5). If method is set
+        it overrides the value of gamma.
 
-        p_i = \frac{w_i}{\sum_j w_j}
+        Channel capacity is determined by
 
-        N = e^{-H}
+        .. math::
+
+            H = \sum_i p_i \ln p_i
+
+            p_i = \frac{w_i}{\sum_j w_j}
+
+            N = e^{-H}
+
+        Kish by
+
+        .. math::
+
+            N = \frac{(\sum_i w_i)^2}{\sum_i w_i^2}
     """
-    with np.errstate(divide='ignore', invalid='ignore'):
-        W = np.array(w)/sum(w)
-        H = np.nansum(np.log(W)*W)
-        return np.exp(-H)
+    if method == 'kish':
+        gamma = 2
+    elif method == 'channel':
+        gamma = 1
+
+    w = w / np.sum(w)
+    if gamma == 1:
+        with np.errstate(divide='ignore', invalid='ignore'):
+            H = np.nansum(np.log(w)*w)
+            return np.exp(-H)
```

Comment:
When I use the `entropy` function the following test in `test_utils.py` fails

```python

def test_sample_compression_1d():
    np.random.seed(0)
    N = 10000
    x_ = np.random.rand(N)
    w_ = np.random.rand(N)
    n = 1000
    x, w = sample_compression_1d(x_, w_, n)
    assert len(x) == n
    assert len(w) == n
    assert np.isclose(w.sum(), w_.sum())

    x_ = np.random.rand(N)
    w_ = np.random.randn(N)
    x, w = sample_compression_1d(x_, w_, ncompress=True)
    assert len(w) <= N

```
Specifically, this bit

```python
x_ = np.random.rand(N)
w_ = np.random.randn(N)
x, w = sample_compression_1d(x_, w_, ncompress=True)
assert len(w) <= N
```
because the `entropy(w)` is evaluated as `-inf` giving 0.0 effective samples and throwing the following error 

```TypeError: expected a sequence of integers or a single integer, got '0.0'```

in the `np.choice` function in `sample_compression_1d()` in `utils.py`.

Reply:
Looks like we've got a bug in `test_sample_compression_1d`. 

The line
```python
    w_ = np.random.randn(N)
```
produces negative weights, but our weights (although not necessarily normalised) should be strictly positive...

Reply:
How about replacing it with

`w_ = np.random.lognormal(1, 0.5, N)`

Reply:
Hmm I get a similar error

`TypeError: expected a sequence of integers or a single integer, got '8828.14997708517'`

Should `ncompress` be an integer?